src/metabase/util/malli/registry.cljc

```diff
@@ -3,28 +3,123 @@
+           :else x))
+       x)))
+
+;;;; `with-key` helpers
+
+(defn- stable-sym-names
+  "Returns a lazy seq of symbols that can be used as stable argument names."
+  []
+  (letfn [(int-to-stable-arg [n]
+            (loop [num n
+                   result ""]
+              (if (zero? num)
+                (symbol result)
+                (let [remainder (mod (dec num) 26)
+                      char (char (+ (int \a) remainder))]
+                  (recur (quot (dec num) 26)
+                         (str char result))))))]
+    (map int-to-stable-arg (iterate inc 1))))
+
+(defn- stabilize-reader-fn-args
+  "Calling with-key on a reader-macro generated function will cause it to have
+  gensym'd arguments
+
+  Using this gives us this property in with-key:
+
+  (= (::key (meta (with-key #(+ 1 %))))
+     (::key (meta (with-key #(+ 1 %)))))
+  ;; => true
+  "
+  [form]
+  (let [arguments (remove
+                   ;; do not replace the `&` argument, since it is used to
+                   ;; indicate that the function takes a varargs
+                   #{'&} (second form))
+        argument-replacement (zipmap arguments (stable-sym-names))]
+    (walk/postwalk
+     (fn [x] (if-let [new (argument-replacement x)] new x))
+     form)))
+
+(defn- fn-form-generated-by-reader-macro? [form]
+  (and (coll? form)
+       (= 'fn* (first form))))
+
+(defn- stabilize-schema-fn-args
+  [form]
+  (walk/postwalk
+   (fn [frm] (if (fn-form-generated-by-reader-macro? frm)
+               (stabilize-reader-fn-args frm)
+               frm))
+   form))
+
+(defn- enforce-with-key-idempotency
+  "Wrapping a schema with [[with-key]] multiple times should be idempotent, so this function
+  removes extra wrapping(s) from a schema form, so that it can be used as a cache key."
+  [form]
+  (let [unwrap-forms #{'with-key 'mr/with-key 'metabase.util.malli.registry/with-key}]
+    (loop [frm form]
+      (if (and (coll? frm)
+               (unwrap-forms (first frm)))
+        (recur (second frm))
+        frm))))
+
+(defmacro with-key
+  "Adds `::mr/key` metadata, which is a pr-str'd string of body, to body. Be careful not to call this
+  on functions taking parameters, since it uses the shape of the literal body passed to it as a key.
+  e.g.:
+    (defn my-schema [] :int)
+    (mr/with-key (my-schema))
+    If you change `my-schema` to return `:keyword` here, the cache will not invalidate properly."
+  [body]
+  `(try (with-meta ~body
+                   (assoc (meta ~body) ::key ~(-> body
+                                                  enforce-with-key-idempotency
+                                                  stabilize-schema-fn-args
+                                                  pr-str)))
+        (catch Exception e#
+          (when (or config/is-dev? config/is-test?)
+            (throw (ex-info (str "Error in with-key:" (pr-str ~body))
+                            {:body ~body
+                             :error e#})))
+          ~body)))
```

Comment:
It's a bad sign that you expand `~body` multiple times. If it happens to contain side effects, there will be multiples of them. You probably want to `let` it once and then reuse.